sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/stream/BufferingStreamObserver.java

```diff
@@ -105,10 +112,10 @@ public void onError(Throwable t) {
         // We check to see if we were able to successfully insert the poison pill at the front of
         // the queue to cancel the processing thread eagerly or if the processing thread is done.
         try {
-          // We shouldn't attempt to insert into the queue if the queue drainer thread is done
-          // since the queue may be full and nothing will be emptying it.
-          while (!queueDrainer.isDone()
-              && !queue.offerFirst((T) POISON_PILL, 60, TimeUnit.SECONDS)) {
+          // The order of these checks is important because short circuiting will cause us to
+          // insert into the queue first and only if it fails do we check that the thread is done.
+          while (!queue.offerFirst((T) POISON_PILL, 60, TimeUnit.SECONDS)
```

Comment:
Why is the order of these checks changing and also conversion to || instead of &&?

Based upon the current logic its going to be insert into the beginning of the queue until the queue drainer thread is done. The only way this loop will exit will be if we were able to insert into the queue and the drainer thread is done.

Ditto on below with onCompleted()

Reply:
Reverted. I don't know how this change arrived.